sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.cu

```diff
@@ -0,0 +1,471 @@
+
+  atomicAdd(&(centroids[voxel_index].x), x);
+  atomicAdd(&(centroids[voxel_index].y), y);
+  atomicAdd(&(centroids[voxel_index].z), z);
+  atomicAdd(&(centroids[voxel_index].i), static_cast<float>(i));
+  atomicAdd(&(centroids[voxel_index].count), static_cast<unsigned int>(1));
+}
+
+__global__ void packCentroidKernel(
+  const CudaVoxelGridDownsampleFilter::Centroid * __restrict__ centroids,
+  const size_t num_valid_voxel, const size_t output_point_step, uint8_t * __restrict__ output,
+  uint8_t * __restrict__ return_type_field, uint16_t * __restrict__ channel_field)
+{
+  size_t index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index >= num_valid_voxel) {
+    return;
+  }
+
+  using output_type = typename CudaVoxelGridDownsampleFilter::OutputPointType;
+  auto dst = getElementPointer<output_type>(output, index, output_point_step, 0);
+
+  dst->x = static_cast<decltype(output_type::x)>(centroids[index].x / centroids[index].count);
+  dst->y = static_cast<decltype(output_type::y)>(centroids[index].y / centroids[index].count);
+  dst->z = static_cast<decltype(output_type::z)>(centroids[index].z / centroids[index].count);
+  dst->intensity =
+    static_cast<decltype(output_type::intensity)>(centroids[index].i / centroids[index].count);
+  // Since `return_type` and `channel` fields cannot be accumulated, fill the representative values
+  // for `return_type` and `channel` fields if they are available in the input pointcloud.
+  dst->return_type =
+    voxel_info_dev.input_return_type_offset.is_valid ? *return_type_field : 0;  // 0: UNKNOWN
+  dst->channel =
+    voxel_info_dev.input_channel_offset.is_valid ? *channel_field : 0;  // dummy laser channel ID
+}
+}  // namespace
+
+CudaVoxelGridDownsampleFilter::CudaVoxelGridDownsampleFilter(
+  const float voxel_size_x, const float voxel_size_y, const float voxel_size_z)
+{
+  voxel_info_.voxel_size.x = voxel_size_x;
+  voxel_info_.voxel_size.y = voxel_size_y;
+  voxel_info_.voxel_size.z = voxel_size_z;
+
+  voxel_info_.output_offsets[0] = 0;
+  voxel_info_.output_offsets[1] = voxel_info_.output_offsets[0] + sizeof(OutputPointType::x);
+  voxel_info_.output_offsets[2] = voxel_info_.output_offsets[1] + sizeof(OutputPointType::y);
+  voxel_info_.output_offsets[3] = voxel_info_.output_offsets[2] + sizeof(OutputPointType::z);
+  voxel_info_.output_offsets[4] =
+    voxel_info_.output_offsets[3] + sizeof(OutputPointType::intensity);
+  voxel_info_.output_offsets[5] =
+    voxel_info_.output_offsets[4] + sizeof(OutputPointType::return_type);
+
+  CUDA_VGDF_ERROR_CHECK(cudaStreamCreate(&stream_));
+
+  // create memory pool to make repeated allocation efficient
+  {
+    int current_device_id = 0;
+    CUDA_VGDF_ERROR_CHECK(cudaGetDevice(&current_device_id));
+    CUDA_VGDF_ERROR_CHECK(cudaDeviceGetDefaultMemPool(&mem_pool_, current_device_id));
+
+    // Configure memory pool reusing allocation
+    // Following CUDA sample, set high release threshold so that allocated memory region
```

Comment:
```suggestion
    // Configure the memory pool reusing allocation
    // Following the CUDA documentation, we set a high release threshold so that the allocated memory region
```